gemini/sample-apps/llamaindex-rag/src/app/main.py

```diff
@@ -0,0 +1,484 @@
+
+
+class IndexUpdate(BaseModel):
+    base_index_name: str
+    base_endpoint_name: str
+    qa_index_name: Optional[str]
+    qa_endpoint_name: Optional[str]
+    firestore_db_name: Optional[str]
+    firestore_namespace: Optional[str]
+
+
+class PromptUpdate(BaseModel):
+    prompt_name: str
+    new_content: str
+
+
+class RAGConfig(BaseModel):
+    llm_name: str = "gemini-1.5-flash"
+    temperature: float = 0.2
+    similarity_top_k: int = 5
+    retrieval_strategy: str = "auto_merging"
+    use_hyde: bool = True
+    use_refine: bool = True
+    use_node_rerank: bool = True
+    use_react: bool = True
+    qa_followup: bool = True
+    hybrid_retrieval: bool = True
+
+
+class RAGRequest(RAGConfig):
+    query: str = "What were Google's Q1 Earnings?"
+    evaluate_response: bool
+    eval_model_name: Optional[str] = "gemini-1.5-flash"
+    embedding_model_name: Optional[str] = "text-embedding-004"
+
+
+class EvalRequest(RAGConfig):
+    eval_model_name: str = "gemini-1.5-flash"
+    embedding_model_name: Optional[str] = "text-embedding-004"
+    input_eval_dataset_bucket_uri: str = "test_rag_questions/test_ground_truth.csv"
+    bq_eval_results_table_id: str = "eval_results.eval_results_table"
+    ragas_metrics: List[str] = ["faithfulness", "answer_relevancy"]
+
+
+# Shared state - to be updated by UI
+from src.app.shared_state import index_manager, prompts
+
+
+def get_index_manager():
+    logger.info(index_manager.base_index.docstore)
+    logger.info(index_manager.firestore_db_name)
+    logger.info(index_manager.firestore_namespace)
+    return index_manager
+
+
+def get_prompts():
+    return prompts
+
+
+creds = google.auth.default()[0]
+request = google.auth.transport.requests.Request()
+
+app = FastAPI()
+
+
+@app.get("/")
+def root():
+    return {"hello world": "hello world"}
+
+
+@app.get("/get_all_prompts")
+async def get_all_prompts(prompts=Depends(get_prompts)):
+    return prompts.to_dict()
+
+
+@app.post("/update_prompt")
+async def update_prompt(prompt_update: PromptUpdate, prompts=Depends(get_prompts)):
+    try:
+        prompts.update(prompt_update.prompt_name, prompt_update.new_content)
+        return {"message": f"Prompt {prompt_update.prompt_name} updated successfully"}
+    except ValueError as e:
+        raise HTTPException(status_code=400, detail=str(e))
+
+
+@app.post("/list_vector_search_indexes")
+async def list_vector_search_indexes(
+    qa_or_base: dict, index_manager=Depends(get_index_manager)
+):
+    index_list = aiplatform.MatchingEngineIndex.list()
+    index_list = [i.display_name for i in index_list]
+    if qa_or_base["qa_or_base"] == "qa":
+        active_index_name = index_manager.qa_index_name
+    elif qa_or_base["qa_or_base"] == "base":
+        active_index_name = index_manager.base_index_name
+    if active_index_name is None:
+        index_list.insert(0, None)
+    else:
+        index_list.remove(active_index_name)
+        index_list.insert(0, active_index_name)
+    return index_list
```

Comment:
I think a lot of this logic could be simplified. Maybe try sending it to Gemini 1.5

Reply:
fixed in [commit](https://github.com/GoogleCloudPlatform/generative-ai/commit/a83634e9ac49061cb5e733cb449f065055638744)